src/Raven.Client/Documents/Session/DocumentSession.Attachments.cs

```diff
@@ -28,5 +30,33 @@ public AttachmentResult GetRevisionAttachment(string documentId, string name, Ch
             var operation = new GetAttachmentOperation(documentId, name, stream, AttachmentType.Revision, changeVector);
             return DocumentStore.Operations.Send(operation);
         }
+
+        public void StoreAttachment(string documentId, string name, Stream stream, string contentType = null)
+        {
+            if (string.IsNullOrWhiteSpace(documentId))
+                throw new ArgumentNullException(nameof(documentId));
+            if (string.IsNullOrWhiteSpace(name))
+                throw new ArgumentNullException(nameof(name));
+
+            if (_deferedCommands.OfType<DeleteCommandData>().Any(c => c.Key == documentId))
+                throw new InvalidOperationException($"Can't store attachment {name} on document {documentId}, there is a deferred command registered for this document to be deleted.");
+
+            if (_deferedCommands.OfType<PutAttachmentCommandData>().Any(c => c.Key == documentId && c.Name == name))
+                throw new InvalidOperationException($"Can't store attachment {name} on document {documentId}, there is a deferred command registered for this document to be deleted.");
+
+            if (DocumentsById.TryGetValue(documentId, out DocumentInfo documentInfo) &&
+                DeletedEntities.Contains(documentInfo.Entity))
+                throw new InvalidOperationException($"Can't store attachment {name} on document {documentId}, the document was already deleted in this session.");
+
+            Defer(new PutAttachmentCommandData(documentId, name, stream, contentType, null));
+        }
+
+        public void StoreAttachment(object entity, string name, Stream stream, string contentType = null)
+        {
+            if (DocumentsByEntity.TryGetValue(entity, out DocumentInfo document) == false)
+                throw new ArgumentException("Entity is not tracked in the seesion. Use documentId instead or track the entity in the session.", nameof(entity));
```

Comment:
Include the type of the entity here, to make it easier to figure out what is going on in the error